File: packages/java/anip-crypto/src/main/java/dev/anip/crypto/JwksSerializer.java
package dev.anip.crypto;

import com.nimbusds.jose.jwk.ECKey;
import com.nimbusds.jose.jwk.JWKSet;
import com.nimbusds.jose.jwk.KeyUse;

import java.util.List;
import java.util.Map;

/**
 * Serializes public keys as a JWKS (JSON Web Key Set).
 */
public final class JwksSerializer {

    private JwksSerializer() {}

    /**
     * Returns the JWKS representation of both public keys.
     * The delegation key has use="sig" and the audit key has use="audit".
     *
     * @param km the key manager
     * @return a Map with the standard JWKS format {"keys": [...]}
     */
    @SuppressWarnings("unchecked")
    public static Map<String, Object> toJwks(KeyManager km) {
        ECKey delegationPublic = new ECKey.Builder(km.getDelegationECKey().toPublicJWK())
                .keyID(km.getDelegationKid())
                .keyUse(KeyUse.SIGNATURE)
                .algorithm(com.nimbusds.jose.JWSAlgorithm.ES256)
                .build();

        // For audit key, we need to set a custom "use" value
        ECKey auditPublic = new ECKey.Builder(km.getAuditECKey().toPublicJWK())
                .keyID(km.getAuditKid())
                .algorithm(com.nimbusds.jose.JWSAlgorithm.ES256)
                .build();

        // Build manually to support custom "use" value for audit key
        Map<String, Object> delMap = delegationPublic.toJSONObject();
        delMap.put("use", "sig");

        Map<String, Object> auditMap = auditPublic.toJSONObject();
        auditMap.put("use", "audit");

        return Map.of("keys", List.of(delMap, auditMap));
    }
}

File: packages/csharp/src/Anip.Crypto/JwksSerializer.cs
using System.Security.Cryptography;

namespace Anip.Crypto;

/// <summary>
/// Serializes public keys as a JWKS (JSON Web Key Set).
/// </summary>
public static class JwksSerializer
{
    /// <summary>
    /// Serializes both public keys from the KeyManager as a JWKS.
    /// Returns a dictionary suitable for JSON serialization as {"keys": [...]}.
    /// Only includes public key parameters (no private key material).
    /// </summary>
    public static Dictionary<string, object> ToJwks(KeyManager keyManager)
    {
        return new Dictionary<string, object>
        {
            ["keys"] = new List<Dictionary<string, object>>
            {
                PublicKeyToJwkMap(keyManager.GetSigningKey(), keyManager.GetSigningKid(), "sig"),
                PublicKeyToJwkMap(keyManager.GetAuditKey(), keyManager.GetAuditKid(), "audit")
            }
        };
    }

    private static Dictionary<string, object> PublicKeyToJwkMap(ECDsa key, string kid, string use)
    {
        var jwk = KeyManager.ExportPublicKeyToJwk(key);
        return new Dictionary<string, object>
        {
            ["kty"] = jwk["kty"],
            ["crv"] = jwk["crv"],
            ["x"] = jwk["x"],
            ["y"] = jwk["y"],
            ["kid"] = kid,
            ["alg"] = "ES256",
            ["use"] = use
        };
    }
}
